sunpy/coordinates/transformations.py

```diff
@@ -187,71 +221,59 @@ def hcc_to_hgs(helioccoord, heliogframe):
                            "heliographic coordinates for observer '{}' "
                            "without `obstime` being specified.".format(helioccoord.observer))
 
-    x = helioccoord.x.to(u.m)
-    y = helioccoord.y.to(u.m)
-    z = helioccoord.z.to(u.m)
-
-    l0_rad = helioccoord.observer.lon
-    b0_deg = helioccoord.observer.lat
+    # Make sure the observer is in HGS
+    obs = helioccoord.observer.\
+          transform_to(HeliographicStonyhurst(obstime=helioccoord.observer.obstime))
 
-    cosb = np.cos(np.deg2rad(b0_deg))
-    sinb = np.sin(np.deg2rad(b0_deg))
+    total_matrix = _rotation_matrix_hcc_to_hgs(obs.lon, obs.lat)
 
-    hecr = np.sqrt(x**2 + y**2 + z**2)
-    hgln = np.arctan2(x, z * cosb - y * sinb) + l0_rad
-    hglt = np.arcsin((y * cosb + z * sinb) / hecr)
+    # Transform from HCC to HGS at the HCC obstime
+    newrepr = helioccoord.cartesian.transform(total_matrix)
+    int_coord = HeliographicStonyhurst(newrepr, obstime=helioccoord.obstime)
 
-    representation = SphericalRepresentation(
-        np.rad2deg(hgln), np.rad2deg(hglt), hecr.to(u.km))
-    return heliogframe.realize_frame(representation)
+    # Loopback transform HGS if there is a change in obstime
+    if np.any(helioccoord.obstime != heliogframe.obstime):
+        return _transform_obstime(int_coord, heliogframe)
+    else:
+        return int_coord
 
 
-@frame_transform_graph.transform(FunctionTransform, HeliographicStonyhurst,
-                                 Heliocentric)
+@frame_transform_graph.transform(FunctionTransformWithFiniteDifference,
+                                 HeliographicStonyhurst, Heliocentric)
 def hgs_to_hcc(heliogcoord, heliocframe):
     """
     Convert from Heliographic Stonyhurst to Heliocentric Cartesian.
     """
-    hglon = heliogcoord.spherical.lon
-    hglat = heliogcoord.spherical.lat
-    r = heliogcoord.spherical.distance
-    if r.unit is u.one and u.allclose(r, 1*u.one):
-        r = np.ones_like(r)
-        r *= RSUN_METERS
-
     if not isinstance(heliocframe.observer, BaseCoordinateFrame):
         raise ConvertError("Cannot transform heliographic coordinates to "
                            "heliocentric coordinates for observer '{}' "
                            "without `obstime` being specified.".format(heliocframe.observer))
 
-    l0_rad = heliocframe.observer.lon.to(u.rad)
-    b0_deg = heliocframe.observer.lat
-
-    lon = np.deg2rad(hglon)
-    lat = np.deg2rad(hglat)
+    # Make sure the observer is in HGS
+    obs = heliocframe.observer.\
+          transform_to(HeliographicStonyhurst(obstime=heliocframe.observer.obstime))
 
-    cosb = np.cos(b0_deg.to(u.rad))
-    sinb = np.sin(b0_deg.to(u.rad))
-
-    lon = lon - l0_rad
-
-    cosx = np.cos(lon)
-    sinx = np.sin(lon)
-    cosy = np.cos(lat)
-    siny = np.sin(lat)
+    # If HGS coord is only directional, place it on the surface of the Sun
```

Comment:
I don't think this is the place we should be doing this, we should be using `calculate_distance` (renamed) to do this.

Reply:
I'm going to remove this code because the HGS docstring says that it will always be 3D.  (In actuality, it can be only directional in some circumstances, but I'll consider that a bug to be fixed.)